src/openms_gui/source/VISUAL/Spectrum2DCanvas.cpp

```diff
@@ -326,100 +326,85 @@ namespace OpenMS
       {
-        if (it->getMSLevel() == 1)
+        if (it->getMSLevel() == 1 && it->size() > 1)
         {
-          ++n_ms1_scans;
+          rt_indices.push_back(std::distance(peak_map.begin(), it));
         }
       }
+      Size n_ms1_scans = rt_indices.size();
 
-      // create iterator on scan in the middle of the visible map
-      Size n_peaks_in_middle_scan(0);
-      ExperimentType::ConstIterator it = peak_map.RTBegin(rt_min) + n_ms1_scans / 2;
-      if (it != peak_map.end())
+      // sample #of points at 3 scan locations (25%, 50%, 75%)
+      // and take the median value
+      Size n_peaks_in_scan(0);
       {
-        for (ExperimentType::SpectrumType::ConstIterator it2 = it->MZBegin(mz_min); it2 != it->end() && it2 != it->MZEnd(mz_max); ++it2)
+        double quantiles[] = {0.25, 0.50, 0.75};
+        std::vector<Size> n_s;
+        for (Size i=0; i<sizeof(quantiles)/sizeof(double); ++i)
         {
-          ++n_peaks_in_middle_scan;
-        }
+          const ExperimentType::SpectrumType& spec = peak_map[rt_indices[n_ms1_scans*quantiles[i]]];
+          n_s.push_back(std::distance(spec.MZBegin(mz_min), spec.MZEnd(mz_max)) + 1); // +1 to since distance is 0 if only one m/z is shown
+        } 
+        std::sort(n_s.begin(), n_s.end());
+        n_peaks_in_scan = n_s[1]; // median
       }
-
-      // determine spacing for whole data
-      double min_spacing_mz = 1.0;
-      double average_spacing_rt = 1.0;
-      Size n_scans = peak_map.size();
-      {
-        vector<float> mz_spacing;
-        for (Size i = 0; i != n_scans; ++i)
-        {
-          // skip non MS1 and empty spectra
-          Size ms_level = peak_map[i].getMSLevel();
-          Size n_peaks = peak_map[i].size();
-          if (ms_level != 1 || n_peaks < 2)
-          {
-            continue;
-          }
-          double current_mz_spacing =  (peak_map[i][n_peaks - 1].getMZ() - peak_map[i][0].getMZ()) / n_peaks;
-          mz_spacing.push_back(current_mz_spacing);
-        }
-        sort(mz_spacing.begin(), mz_spacing.end());
-        min_spacing_mz = !mz_spacing.empty() ? mz_spacing[0] : 1.0;
+      
+      double ratio_data2pixel_rt = n_ms1_scans / (double)rt_pixel_count;
+      double ratio_data2pixel_mz = n_peaks_in_scan / (double)mz_pixel_count;
 
 #ifdef DEBUG_TOPPVIEW
-        cout << "BEGIN " << __PRETTY_FUNCTION__ << endl;
-        cout << "min spacing mz:" << min_spacing_mz << endl;
+      std::cerr << rt_min << ":" << rt_max <<"   " << mz_min << ":" << mz_max << "\n";
+      std::cerr << n_ms1_scans << "rt " << n_peaks_in_scan << "ms\n";
+      std::cerr << rt_pixel_count << " x " << mz_pixel_count << " px\n";
+      std::cerr << ratio_data2pixel_rt << " " << ratio_data2pixel_mz << " ratio\n";
 #endif
+      // minimum fraction of image expected to be filled with data
+      // if not reached, we upscale point size
+      const double MIN_COVERAGE = 0.2;
```

Comment:
seems like a magic constant that we might even want to adopt at some point.  can you move this into the header and either make it a preprocessor macro or static variable inside the OpenMS namespace (preferred option), maybe even inside its own namespace `OpenMS::VISUAL::MIN_COVERAGE`

this is a minor point but maybe makes sense